### ISSUE TITLE
FormFlow loses context in all channels except emulator

### ISSUE THREAD

**Issue by DrWala** [microsoft/BotBuilder-V3#137]
## Version
3.16.1

## Describe the bug
Formflow typically works by asking the user a set of questions. In the emulator, formflow works as expected. In webchat, directline or telegram, answering the question posed by formflow causes the convo to restart back to the rootcontroller.

This bot was working fine for a while, and today pushing an update broke all the channels except the emulator

## Expected behavior
![image](https://user-images.githubusercontent.com/6803997/56955851-12aa3100-6b75-11e9-97f3-e8b08f0fda7b.png)
Bot asks formflow questions in order (correct)

![image](https://user-images.githubusercontent.com/6803997/56955909-3a00fe00-6b75-11e9-828e-a335aaeba631.png)
Bot resets after an answer is sent (telegram)

![image](https://user-images.githubusercontent.com/6803997/56955947-4d13ce00-6b75-11e9-8939-d8d4dd85a8b7.png)
Bot resets after an answer is sent (webchat)

Would appreciate assistance asap as this is a production bot.

**Comment by EricDahlvang** [comment]
Have you tried debugging webchat or directline locally?  https://blog.botframework.com/2017/10/19/debug-channel-locally-using-ngrok/

**Comment by DrWala** [comment]
I haven't done that yet but I'll give it a shot. Some further info based on my testing:

The formflow that's being called is the result of a proactive dialog being fired off after a time interval through hangfire. This previously was working fine until I pushed a completely unrelated change and it resulted in this specific feature failing.

I tested a formflow that's not fired through a proactive dialog and it runs fine. It seems the proactive dialog specifically seems to not allow the formflow to finish, ending immediately after the first question. I will test using the ngrok based local debugging and update you guys. 

Thanks for your assistance!

**Comment by EricDahlvang** [comment]
Can you share the code you're using to send the proactive message?

**Comment by DrWala** [comment]
Here is the code I got off the documentation here: https://docs.microsoft.com/en-us/azure/bot-service/dotnet/bot-builder-dotnet-proactive-messages?view=azure-bot-service-3.0

```
using (var scope = DialogModule.BeginLifetimeScope(Conversation.Container, activity))
{
    var botData = scope.Resolve<IBotData>();
    await botData.LoadAsync(CancellationToken.None);

    // This is the dialog stack.
    var stack = scope.Resolve<IDialogTask>();

    // Create the new dialog and add it to the stack.
    var dialog = new CallProcedureDialog(procedure: procedure, patient: patient);
    //I've optionally tried doing this but it didn't do much
    //stack.Reset();
    stack.Call(dialog.Void<object, IMessageActivity>(), null);
    await stack.PollAsync(CancellationToken.None);

    // Flush the dialog stack back to its state store.
    await botData.FlushAsync(CancellationToken.None);
}
```

**Comment by EricDahlvang** [comment]
What storage prover is this bot using for state?

**Comment by DrWala** [comment]
I think at the moment it's using the in memory storage, which I should probably change...

**Comment by EricDahlvang** [comment]
@DrWala it looks like this line should be changed:
```cs
Activity activity = JsonConvert.DeserializeObject<Activity>(patient.BotUserData.ConversationReference);
```
to this:
```cs
var activity = JsonConvert.DeserializeObject<ConversationReference>(patient.BotUserData.ConversationReference).GetPostToBotMessage();
```

A ConversationReference does not deserialize to an Activity, but rather to a ConversationReference.

**Comment by EricDahlvang** [comment]
@DrWala  Closing this.  Please let us know if you are still working on it, and we can re-open the issue.

**Comment by DrWala** [comment]
Hi @EricDahlvang ,

Sorry, been under the weather the past 2 weeks and haven't touched work. I'm trying to do this now but it seems its not a method that exists in the first place.

![Untitled-1](https://user-images.githubusercontent.com/6803997/58757111-8bb3f400-8538-11e9-98ff-2d57647aa594.png)

Am I using the wrong SDK or something?

Bot Builder is on 3.16
![image](https://user-images.githubusercontent.com/6803997/58757127-c74ebe00-8538-11e9-9c1c-76d83bdcb6cd.png)

I've also switched to table storage, if thats useful info for you.

**Comment by EricDahlvang** [comment]
Hi @DrWala

The ConversationReference should be deserialized to a ConversationReference, not to an Activity.

'''cs
var activity = JsonConvert.DeserializeObject<ConversationReference>(patient.BotUserData.ConversationReference).GetPostToBotMessage()

'''

**Comment by DrWala** [comment]
Hi @EricDahlvang 

It seems to throw an exception - Object reference not set to instance of an object. However, previously treating it as an activity had no issues so I'm not sure what's going on.

I've attached the code for how I save the user's activity as well if this helps.

```
currentPatient.BotUserData = new BotUserData()
{
    ToId = context.Activity.Recipient.Id,
    ToName = context.Activity.Recipient.Name,
    FromId = context.Activity.From.Id,
    FromName = context.Activity.From.Name,
    ServiceUrl = context.Activity.ServiceUrl,
    ConversationId = context.Activity.Conversation.Id,
    ChannelId = context.Activity.ChannelId,
    ConversationReference = JsonConvert.SerializeObject(context.Activity.AsMessageActivity())
};
```

**Comment by EricDahlvang** [comment]
Please try using `Activity.ToConversationReference()`

```cs
currentPatient.BotUserData = new BotUserData()
                {
                    ToId = context.Activity.Recipient.Id,
                    ToName = context.Activity.Recipient.Name,
                    FromId = context.Activity.From.Id,
                    FromName = context.Activity.From.Name,
                    ServiceUrl = context.Activity.ServiceUrl,
                    ConversationId = context.Activity.Conversation.Id,
                    ChannelId = context.Activity.ChannelId,
                    ConversationReference = JsonConvert.SerializeObject(context.Activity.ToConversationReference())
                };
```

**Comment by DrWala** [comment]
That's not a valid method apparently. However.... it seems I somehow got it working and I have no idea what I did. I think the only thing I changed was:

`private async Task MessageReceivedAsync(IDialogContext context, IAwaitable<IMessageActivity> result)`

whereas it previously was

`private async Task MessageReceivedAsync(IDialogContext context, IAwaitable<object> result)`

but I really don't think it has an impact?
